pkg/globalstats/stats.go

```diff
@@ -1,168 +1,248 @@
-	engine, ok := lazyInitializeEngine()
-	if !ok {
-		return
+func Incr(name, family, table string) {
+	k := counterKey{name: name, family: family, table: table}
+	select {
+	case eventChan <- statEvent{typ: statEventTypeIncr, incr: k}:
+	default:
+		// eventChan is full, drop this stat
+		incrDroppedStats()
 	}
-
-	engine.Incr(name, tags...)
 }
 
-func Disable() {
-	mut.Lock()
-	defer mut.Unlock()
-
-	stopped = true
-
-	// Stop any goroutines launched from any previous Initialize calls.
-	if flusherStop != nil {
-		close(flusherStop)
-		flusherStop = nil
+func Observe(name string, value interface{}, tags ...stats.Tag) {
+	k := observation{name: name, value: value, tags: tags}
+	select {
+	case eventChan <- statEvent{typ: statEventTypeObserve, observe: k}:
+	default:
+		// eventChan is full, drop this stat
+		incrDroppedStats()
 	}
 }
 
+// Initialize configures ctlstore to report global stats. This must be called
+// at least once in order to report stats.
 func Initialize(ctx context.Context, cfg Config) {
-	mut.Lock()
-	defer mut.Unlock()
-
-	config = cfg
-	globalctx = ctx
-}
-
-func lazyInitializeEngine() (*stats.Engine, bool) {
-	if stopped || engine != nil {
-		return engine, !stopped
-	}
-
-	mut.Lock()
-	defer mut.Unlock()
-
-	// Perform a second check after grabbing the mutex, in case there was competition for the lock.
-	if stopped || engine != nil {
-		return engine, !stopped
-	}
-
-	if config.AppName == "" {
+	// Set any default configuration values, where necessary.
+	if cfg.AppName == "" {
 		if len(os.Args) > 0 {
-			config.AppName = filepath.Base(os.Args[0])
+			cfg.AppName = filepath.Base(os.Args[0])
 		} else {
-			config.AppName = "unknown"
+			cfg.AppName = "unknown"
 		}
 	}
-	if config.SamplePct == 0 {
+	if cfg.SamplePct == 0 {
 		// By default, only sample 10% of the observations.
-		config.SamplePct = 0.10
-	}
-	if config.FlushEvery == 0 {
-		config.FlushEvery = 10 * time.Second
+		cfg.SamplePct = 0.10
 	}
-	if config.StatsHandler == nil {
-		config.StatsHandler = stats.DefaultEngine.Handler
+	if cfg.FlushEvery == 0 {
+		cfg.FlushEvery = 10 * time.Second
 	}
-	if config.CtlstoreVersion == "" {
-		config.CtlstoreVersion = version
+	if cfg.StatsHandler == nil {
+		cfg.StatsHandler = stats.DefaultEngine.Handler
 	}
-	if globalctx == nil {
-		globalctx = context.Background()
+	cfg.ctx = ctx
+	if cfg.ctx == nil {
+		cfg.ctx = context.Background()
 	}
 
-	err := func() error {
-		if config.SamplePct > 1 || config.SamplePct < 0 {
-			return errors.New("sample percentage must be in the range of (0, 1]")
-		}
-		if config.FlushEvery < 0 {
-			return errors.New("flush rate must be a positive duration")
-		}
-
-		return nil
-	}()
-	if err != nil {
-		events.Log("Could not initialize ctlstore globalstats: %{error}s", err)
-		return nil, false
+	// Validate that all config values are valid.
+	if cfg.FlushEvery < 0 {
+		events.Log("Could not initialize flagon global stats: %{error}s", errors.New("flush rate must be a positive duration"))
```

Comment:
s/flagon/ctlstore

Reply:
Good catch! Thought I grepped for `flagon` before pushing, but I guess not 😂